tools/link.py

```diff
@@ -606,25 +606,88 @@ def include_and_export(name):
       settings.EXPORTED_RUNTIME_METHODS += ['ExitStatus']
 
 
+def set_initial_memory():
+  # INITIAL_HEAP cannot be used when the memory object is created in JS.
+  if settings.IMPORTED_MEMORY:
+    if 'INITIAL_HEAP' not in user_settings:
+      # The default for imported memory is to fall back to INITIAL_MEMORY.
+      settings.INITIAL_HEAP = -1
+    else:
+      # Some of these could (and should) be implemented.
+      exit_with_error('INITIAL_HEAP is currently not compatible with IMPORTED_MEMORY, SHARED_MEMORY, RELOCATABLE, ASYNCIFY_LAZY_LOAD_CODE, WASM2JS')
```

Comment:
1) `IMPORTED_MEMORY` - this is just harmless, but I didn't want to complicate this further trying to detect if `IMPORTED_MEMORY` implicit or not.
2) `SHARED_MEMORY` - ?.
3) `RELOCATABLE` - not sure how the memory scheme here works. Do the side modules import the main module's memory?
4) `ASYNCIFY_LAZY_LOAD_CODE` - ?.
5) `WASM2JS` - apparently it is a code size regression to define memory in WASM in this mode. It is also a legacy mode.

Reply:
This seems like a new error?   Right now I can do `emcc -sINITIAL_MEMORY=32mb -sIMPORTED_MEMORY hello.c` without any error.

Reply:
The error is under `if 'INITIAL_HEAP' in user_settings`, so it should only apply if you try to use `INITIAL_HEAP` with an unsupported (unimplemented) configuration.

Reply:
Right, I'm just confirming that that was not previously an error.

Reply:
How about `INITIAL_HEAP is currently not compatible with imported memoy`?

Also, I think `WASM2JS` should still supported `INITIAL_HEAP` even through the memory is technically imported (since we still pragmatically create the memory with the correct size).   Can you explain why this is not possible?

I wonder if we should switch wasm2js to not use imported memory before we land this change?

Reply:
> How about `INITIAL_HEAP is currently not compatible with imported memory`?

I think it would technically work, but it's a bit misleading. There are two cases:
1) `IMPORTED_MEMORY` is set because the user set it, and they want to provide the memory object themselves. In this case, both `INITIAL_MEMORY` and `INITIAL_HEAP` are basically meaningless. The only reason this scenario is blocked in this change is for simplicity: I didn't want to write code to detect whether `IMPORTED_MEMORY` is set by the user, or implicitly, or both. I presume that explicit `IMPORTED_MEMORY` is rare.
2) `IMPORTED_MEMORY` is set as an internal implementation detail of Emscripten, where it is convenient/necessary to instantiate the memory object in JS. 'Leaking' that into documentation would seem odd.

> Also, I think WASM2JS should still supported INITIAL_HEAP even through the memory is technically imported (since we still pragmatically create the memory with the correct size). Can you explain why this is not possible?

I believe it is possible to switch all `INITIAL_MEMORY` cases to `INITIAL_HEAP` and delete the error. It is just work I have not done yet. For it to be done, I basically need to understand what are the reasons for implicit `IMPORTED_MEMORY` in all the cases that it is set, and what is the optimal solution in each case.

For WASM2JS specifically, it seems like there is a code size optimization in place where doing imported memory "manually" saves code size compared to the code generated by the `wasm2js` tool for WASM modules that define their own memory. So here the fix could be to improve the `wasm2js` tool, or read out the defined memory size from the linked WASM binary (this could be the solution to other cases too), or something else.